core/src/test/scala/org/apache/spark/serializer/JavaSerializerSuite.scala

```diff
@@ -25,4 +25,18 @@ class JavaSerializerSuite extends SparkFunSuite {
     val instance = serializer.newInstance()
     instance.deserialize[JavaSerializer](instance.serialize(serializer))
   }
+
+  test("Deserialize object containing a primitive Class as attribute") {
+    val serializer = new JavaSerializer(new SparkConf())
+    val instance = serializer.newInstance()
+    instance.deserialize[JavaSerializer](instance.serialize(new ContainsPrimitiveClass()))
+  }
```

Comment:
I'd like to have a test that we still get the original ClassNotFound in other cases, but I'm not sure how to do that ... perhaps you could write a test where you serialize with a special classloader which loads an extra class, and then deserialize with a different classloader?